### PR TITLE
Resolve dependencies when updating mods

### PR DESCRIPTION
**Pull request by Trial97**
Integrates dependency checking with the mod update task, resulting in the filling of missing dependencies if any.

### CODE REVIEW

**Comment by getchoo** [comment]
i'm assuming this would partially solve https://github.com/PrismLauncher/PrismLauncher/issues/450?

**Comment by Trial97** [comment]
Yeah the dependency part, not the incompatible part 
But let me tweak this PR a little

**Comment by TheKodeToad** [comment]
this doesn't really look related to the issue
that is all about indicators for mods which won't load

**Comment by Trial97** [comment]
I will not implement the incompatible part, because it looks like curseforge API doesn't really return that information.
For example, https://www.curseforge.com/minecraft/mc-mods/universal-tweaks has some incompatible mods in the platform but when I query only the required dependencies(no other dependency type) are present.
So this PR will not implement the mentioned issue so this PR is ready for review :)

**Comment by getchoo** [comment]
> that is all about indicators for mods which won't load

i was referring to the "show a list of missing mod dependencies" part, since that would be solved here as they are just installed automatically
reading it again though, maybe that is a bit of a stretch :p

**Review by getchoo — COMMENTED** [review]
it works as expected...*but* there was one failed assertion i got when building on msvc. 
image

i'm not exactly sure why this was triggered, and i haven't been able to get it again, even when repeating the same steps of new instance->install fabric->install iris

**Comment by Trial97** [comment]
> it works as expected..._but_ there was one failed assertion i got when building on msvc. image
> 
> i'm not exactly sure why this was triggered, and i haven't been able to get it again, even when repeating the same steps of new instance->install fabric->install iris

This is not a bug created by this PR, if I remember @TayouVR  has encounter the same here:https://discord.com/channels/1031648380885147709/1031823065937629267/1121460300642795613

**Comment by getchoo** [comment]
ah gotcha, i was just confused because i hadn't seen it yet :derp:

everything seems good to go then!

**Review by getchoo — APPROVED** [review]

**Comment by TheKodeToad** [comment]
Maybe `GetModDependenciesTask::PackDependency` should be elsewhere?

**Comment by Trial97** [comment]
Do you have any suggestions? As right now I keep it there in order to have all the dependency resolution stuff in one place

**Comment by Scrumplex** [comment]
I think rebasing this should be easy with just that one file

File: launcher/modplatform/CheckUpdateTask.h [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #pragma once
 
 #include "minecraft/mod/Mod.h"
+
```
Comment (Trial97) [inline launcher/modplatform/CheckUpdateTask.h]:
```suggestion
```

File: launcher/modplatform/CheckUpdateTask.h [hunk 2]
```diff
@@ -1,6 +1,9 @@
 #pragma once
 
 #include "minecraft/mod/Mod.h"
+#include "minecraft/mod/tasks/GetModDependenciesTask.h"
+
```
Comment (Trial97) [inline launcher/modplatform/CheckUpdateTask.h]:
```suggestion
```

**Comment by Trial97** [comment]
~~This should now be able to merge(the extra commits are some minor line removal that I couldn't get with GitHub merge editor)~~
Well not all went according to plan but at least I have this back (and hopefully running) 

~~Also the windows failing doesn't look like is related to the current changes(maybe some bugs with the pipeline)~~

File: launcher/minecraft/mod/tasks/GetModDependenciesTask.cpp [hunk 3]
```diff
@@ -251,3 +251,32 @@ void GetModDependenciesTask::removePack(const QVariant addonId)
             ++it;
 #endif
 }
+
+QHash<QString, QStringList> GetModDependenciesTask::getRequiredBy()
+{
+    QHash<QString, QStringList> rby;
+    auto fullList = m_selected + m_pack_dependencies;
+    for (auto mod : fullList) {
```
Comment (TheKodeToad) [inline launcher/minecraft/mod/tasks/GetModDependenciesTask.cpp]:
```suggestion
    for (auto& mod : fullList) {
```

**Review by TheKodeToad — APPROVED** [review]

**Pull request merged**